pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_index.go

```diff
@@ -0,0 +1,438 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package scbuildstmt
+
+import (
+	"fmt"
+
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgnotice"
+	"github.com/cockroachdb/cockroach/pkg/sql/privilege"
+	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlerrors"
+	"github.com/cockroachdb/errors"
+)
+
+// DropIndex implements DROP INDEX.
+// It resolves the to-be-dropped index into elements and inform `BuildCtx` to drop them.
+// A considerable amount of effort goes to find the table name of the to-be-dropped index,
+// if not specified (e.g. for `DROP INDEX idx`, we need to search all tables in current db
+// and schema that has an index with name `idx`).
+func DropIndex(b BuildCtx, n *tree.DropIndex) {
+	if n.Concurrently {
+		b.EvalCtx().ClientNoticeSender.BufferClientNotice(b,
+			pgnotice.Newf("CONCURRENTLY is not required as all indexes are dropped concurrently"))
+	}
+
+	for _, index := range n.IndexList {
+		toBeDroppedIndexElms := b.ResolveTableIndexBestEffort(index, ResolveParams{
+			IsExistenceOptional: n.IfExists,
+			RequiredPrivilege:   privilege.CREATE,
+		})
+		if toBeDroppedIndexElms == nil {
+			// Attempt to resolve this index failed but `IF EXISTS` is set.
+			b.MarkNameAsNonExistent(&index.Table)
+			continue
+		}
+
+		// Panic if dropping primary index.
+		_, _, pie := scpb.FindPrimaryIndex(toBeDroppedIndexElms)
+		if pie != nil {
+			panic(errors.WithHint(
+				pgerror.Newf(pgcode.FeatureNotSupported, "cannot drop the primary index of a table using DROP INDEX"),
+				"instead, use ALTER TABLE ... ALTER PRIMARY KEY or"+
+					"use DROP CONSTRAINT ... PRIMARY KEY followed by ADD CONSTRAINT ... PRIMARY KEY in a transaction",
+			))
+		}
+
+		// TODO (Xiang): Check if requires CCL binary for eventual zone config removal.
+
+		// Get the to-be-dropped secondary index element and its name element.
+		_, _, sie := scpb.FindSecondaryIndex(toBeDroppedIndexElms)
+		if sie == nil {
+			panic("cannot find secondary index after resolve. Check resolve logic!")
```

Comment:
I think it's probably better to do:

```
panic(errors.AssertionFailedf("programming error: cannot find secondary index for ... after having resolved it", ...))
```
and then put in some more information about the table and index etc. Same for the below errors.